pynvim/__init__.py

```diff
@@ -77,8 +83,30 @@ def start_host(session=None):
     host.start(plugins)
 
 
-def attach(session_type, address=None, port=None,
-           path=None, argv=None, decode=None):
+@overload
+def attach(session_type: Literal['tcp'], address: str, port: int = 7450) -> Nvim: ...
+
+
+@overload
+def attach(session_type: Literal['socket'], *, path: str) -> Nvim: ...
+
+
+@overload
+def attach(session_type: Literal['child'], *, argv: List[str]) -> Nvim: ...
+
+
+@overload
+def attach(session_type: Literal['stdio']) -> Nvim: ...
+
+
+def attach(
+    session_type: TTransportType,
+    address: str = None,
+    port: int = 7450,
```

Comment:
Is having a default argument here OK? Also, `str = None` is not a valid typing annotation. I think they might be `port: Optional[int] = None` or `address: Optional[str] = None`, etc.

Reply:
You are correct that the explicit `Optional[type]` is now the preferred style, [though it was not always such](https://peps.python.org/pep-0484/#union-types), and mypy I think still supports implicitly inferring the `Optional` (it did at the time of writing this PR, but it's been a while).

Reply:
The default argument does seem like a little bit of a weird change to be included in this PR... maybe something that was left in from testing/debugging?